libtransmission/peer-mgr.cc

```diff
@@ -994,6 +995,69 @@ tr_priority_t tr_swarm::WishlistMediator::priority(tr_piece_index_t piece) const
     return tor_.piece_priority(piece);
 }
 
+bool tr_swarm::WishlistMediator::try_hotswap(tr_block_index_t block, tr_peer const* peer)
+{
+    static auto slow_speed_threshold = Speed{ tr_block_info::BlockSize, Speed::Units::Byps };
+    static auto fast_speed_threshold = Speed{ tr_block_info::BlockSize * 3, Speed::Units::Byps };
+
+    if (peer == nullptr)
+    {
+        return false;
+    }
+
+    auto const now_msec = tr_time_msec();
+    auto const peer_speed = peer->get_piece_speed(now_msec, TR_PEER_TO_CLIENT);
+
+    // If peer is very slow, don't bother hotswapping
+    if (peer_speed < slow_speed_threshold)
+    {
+        return false;
+    }
+
+    tr_peerMsgs* victim = nullptr;
+    auto victim_speed = Speed{};
+
+    // Find the slowest peer (victim) that has this requested block
+    for (auto* const p : swarm_.peers)
+    {
+        if (p == peer || !p->active_requests.test(block))
+        {
+            continue;
+        }
+
+        auto const p_speed = p->get_piece_speed(now_msec, TR_PEER_TO_CLIENT);
+        // Don't hotswap if one peer is already fast enough
+        if (p_speed >= fast_speed_threshold)
+        {
+            return false;
+        }
+
+        if (victim == nullptr || p_speed < victim_speed)
+        {
+            victim = p;
+            victim_speed = p_speed;
+        }
+    }
+
+    // If we found a victim and it's significantly slower,
+    // we can hotswap the block to the current peer
+    if (victim != nullptr && peer_speed > (victim_speed * 2))
```

Comment:
Same here as the `3` above.

Reply:
Also I would remove the second condition in a first time. 

```suggestion
    if (victim != nullptr)
```

You would potentially hide interesting log. At least I would be curios to have the slower peer to have an order of magnitude.

Reply:
Yeah unfortunately if we write to logs frequently, the GTK app freezes during my tests in my case :(